### PR TITLE
AMQP-537: RabbitTemplate: receive() with timeout

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/AMQP-537

### CODE REVIEW

**Comment by artembilan** [comment]
I've kept that like two separate commits, but feel free to squash them on merge.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java [hunk 1]
```diff
@@ -765,25 +766,31 @@ else if (isChannelTransacted()) {
...
-				else {
-					delivery = consumer.nextDelivery(receiveTimeout);
-				}
+                if (timeoutMillis < 0) {
+                    delivery = consumer.nextDelivery();
+                }
+                else {
+                    delivery = consumer.nextDelivery(timeoutMillis);
+                }
 				channel.basicCancel(consumer.getConsumerTag());
```
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java]: tabs/spaces - will fix on merge

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java [hunk 2]
```diff
@@ -814,7 +821,17 @@ public Object receiveAndConvert() throws AmqpException {
...
+	}
+
+	@Override
+	public Object receiveAndConvert(long timeoutMillis) throws AmqpException {
+		return receiveAndConvert(this.getRequiredQueue(), timeoutMillis);
+	}
+
+	@Override
+	public Object receiveAndConvert(String queueName, long timeoutMillis) throws AmqpException {
+		Message response = receive(queueName, timeoutMillis);
```
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java]: Failing tests - need to call `doReceiveNoWait` if the timeout is 0 - will fix on merge.

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java]: Doh. I didn't test the entire project with hope to take a look to the Travis afterwards.
But as you see I forgot that :cry: 

Have you fixed everything already?
Or can I push it?

Thank you anyway!

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java]: BUILD SUCCESSFUL

Merging

**Comment by garyrussell** [comment]
Polish here: https://github.com/garyrussell/spring-amqp/commit/bf0df8bb7d74b0b923d016811ccd9ac28e81d647

**Comment by garyrussell** [comment]
Merged as d896321f42d45ecc3b8c7471cce32a113764754a

**Pull request closed without merging**